Nanopore DNA sequencing is portable and relatively cheap, allowing real time sequencing in the field.  We see the potential to use nanopore sequencing as an accessible educational experience. With a clear pipeline that Just Works(TM), a citizen scientist could do WIMP (What's in my pot?) analysis on their own samples without the need for any external tools.  Undergrad or high school students could follow the steps of the pipeline to learn about the basics of genome assembly.

# Why Nanopore is amazing
* Long reads, cheap, portable
* Comparison to the current standard (Illumina)
* Used for detecting ebola/zika?  Sent to space
* Sequencing in the jungle (tweet below!)
* Idea we care about: you should be able to take a random sample of stuff (ocean water?  dirt?) and sequence it cheaply and easily to find out what's there

To see a video of nanopore sequencing in the jungle, click on the below block of code and click the "Run" button at the top of this page.  You can see on the left side of the code block that it's been run x number of times ("In [x]").  You'll see a star while the code is running ("In [*]") and once it's done, you'll be able to see the video.

In [1]:
%%html
<blockquote class="twitter-tweet" data-lang="en"><p lang="en" dir="ltr">Welcome to my laboratory :)<br><br>Sequencing long ribosomal cluster from plants, insects &amp; fungi in real-time in the Amazon rainforest. Within a few mins of <a href="https://twitter.com/nanopore?ref_src=twsrc%5Etfw">@nanopore</a> data generated, performed BLAST &amp; got correct hits! Dual indexing looks great for pooling many samples<a href="https://twitter.com/hashtag/junglegenomics?src=hash&amp;ref_src=twsrc%5Etfw">#junglegenomics</a> <a href="https://t.co/UQVjYfmU8U">pic.twitter.com/UQVjYfmU8U</a></p>&mdash; Aaron Pomerantz (@AaronPomerantz) <a href="https://twitter.com/AaronPomerantz/status/980873273348038656?ref_src=twsrc%5Etfw">April 2, 2018</a></blockquote>
<script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>

# How Nanopore actually works
Insert a good paragraph description here:

My current understanding is just like, DNA strand pulled through a pore, current across pore changes depending on the nucleotide, you end up with an electrical signal in fast5 format that can be converted to a fasta file (could be fastq but no one trusts quality scores?)

## Poretools
The raw data we get from nanopore sequencing is in the fast5 format.  This is just a series of current values that were read across the pore as the DNA strand passed through it.

We are going to start by looking at this fast5 data, containing current values, and converting it to a fasta file that contains nucleotides. 

This poretools tutorial is adapted from here: http://nbviewer.jupyter.org/github/arq5x/poretools/blob/master/poretools/ipynb/test_run_report.ipynb

First we're going to find our fast5 files.  Our sample fast5 file is in the "data" folder, so we set the variable dataDirectory to "data/".  If you are using your own data, change dataDirectory to the path to your .fast5 files.

In [2]:
# dataDirectory is the path to our fast5 file.
# If you are using your own data, change dataDirectory to the path to your .fast5 files.
dataDirectory = 'data/'

# Print the number of fast5 files in the dataDirectory.
# Click the "Run" button at the top of this page to run this code.
!find $dataDirectory -maxdepth 1 -name "*.fast5" | wc -l

2


Poretools has a number of command line options.  Let's start with the stats command, which will give us some basic statistics about our reads.

In [3]:
# The -q option stops poretools outputting any warning messages.
!poretools stats -q $dataDirectory

/usr/local/lib/python2.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
total reads	6
total base pairs	25217
mean	4202.83
median	4205
min	2940
max	5826
N25	5079
N50	5011
N75	3399


Our sample data has 6 reads and 25,217 base pairs. (Anything else of interest to say about this info?)

We have 3 reads per fast5 because forward, reverse, and two-directional reads are all counted separately. (Is this correct?) We can see the stats about each of these types of reads using the below code.

In [4]:
# Look at stats for forward strands
!poretools stats -q --type fwd $dataDirectory

/usr/local/lib/python2.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
total reads	2
total base pairs	8019
mean	4009.50
median	4009
min	2940
max	5079
N25	5079
N50	5079
N75	2940


In [5]:
# Look at stats for reverse strands
!poretools stats -q --type rev $dataDirectory

/usr/local/lib/python2.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
total reads	2
total base pairs	7973
mean	3986.50
median	3986
min	2962
max	5011
N25	5011
N50	5011
N75	2962


In [6]:
# Look at two-directional reads
!poretools stats -q --type 2D $dataDirectory

/usr/local/lib/python2.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
total reads	2
total base pairs	9225
mean	4612.50
median	4612
min	3399
max	5826
N25	5826
N50	5826
N75	3399


Hopefully we are going to add how to make a squiggle plot here at some point, that shows the current changing and gives a good idea of what signal actually looks like.

In [7]:
# Add squiggle plot here!!!

Now we are going to convert our fast5 file to fasta.  Fasta is a common format for storing DNA sequences.  The below code will take each of the fast5 files in dataDirectory, create a fasta file of that sequence, and store it in a folder called fastaOutput.

In [8]:
# Make a folder to store our fasta files in.
!mkdir fastaOutput

# Convert our fast5 files to fasta.
!poretools fasta $dataDirectory > fastaOutput/outputPoretoolData.fasta

mkdir: cannot create directory ‘fastaOutput’: File exists
/usr/local/lib/python2.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


We can look at the first few lines of this fasta file to see what's in it.  Each of the sequences has a line containing ">" and then a unique identfier, followed by a line containing the nucleotide sequence.

In [9]:
# This will show us the first 200 characters of the first two lines of our file.
# We don't want to look at the whole sequence because it's going to be really long!
!cut -c -200 fastaOutput/outputPoretoolData.fasta | head -2

>1405b039-941f-4f83-9a0b-1e746e05a1de_Basecall_2D_2d CPHG_CNU4299G4G_louse_library_2016_3_4_3507_1_ch120_read353_strand data/2016_3_4_3507_1_ch120_read353_strand.fast5
AAGTGTCGTGCTCTATTGCTTTCTTTTTCTTCTTATATGGTTTGCCCACATCTTCTTTGACCAAATTTTCTTCAACGCTTCCTTCAGTTTTGATCCAGCTTAAGAACCACTTCCTATCGATCTGTTAACGTGCTTCTAACTTCATTTTCCAAAGTTTAAAACATTGTCTTCTATTCGCTTCTTTTCTTTGTTGCTAACAC


This fasta file containing our raw reads will be the input to the next steps in our pipeline.

In [10]:
!poretools winner $dataDirectory > winner.fasta

/usr/local/lib/python2.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


# Alignment and assembly

(I don't really know how to explain this so I'll put Brandon's explanation in here)

Minimap2, we take our fasta file of raw reads and we try to align all of the reads to each other (all vs. all).  In order to do this we tell minimap2 to compare our fasta file to itself.  We write this to another file that contains the (alignment?) called overlap.paf.

Because real fast5 files are humongous (very high frequency resolution), we didn't use a real fast5 sample for poretools.  For this step, we want real sample data, so we're actually going to use a different fasta file.  We're going to wget this and then run minimap2 on it.


In [11]:
!wget https://nanopore.s3.climb.ac.uk/MAP006-1_2D_pass.fasta

--2018-04-04 23:41:11--  https://nanopore.s3.climb.ac.uk/MAP006-1_2D_pass.fasta
Resolving nanopore.s3.climb.ac.uk (nanopore.s3.climb.ac.uk)... 137.205.52.6
Connecting to nanopore.s3.climb.ac.uk (nanopore.s3.climb.ac.uk)|137.205.52.6|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 255168226 (243M) [text/plain]
Saving to: ‘MAP006-1_2D_pass.fasta.2’

MAP006-1_2D_pass.fa 100%[===================>] 243.35M  9.54MB/s    in 27s     

2018-04-04 23:41:41 (8.92 MB/s) - ‘MAP006-1_2D_pass.fasta.2’ saved [255168226/255168226]



Now we're going to tell minimap2 to try to align these reads to each other and write the overlaps to a file called overlap.paf.

In [22]:
!minimap2 -x ava-ont -t 1 MAP006-1_2D_pass.fasta MAP006-1_2D_pass.fasta > overlap.paf

[M::mm_idx_gen::9.833*1.00] collected minimizers
[M::mm_idx_gen::16.358*1.00] sorted minimizers
[M::main::16.358*1.00] loaded/built the index for 25483 target sequence(s)
[M::mm_mapopt_update::17.041*1.00] mid_occ = 52
[M::mm_idx_stat] kmer size: 15; skip: 5; is_hpc: 0; #seq: 25483
[M::mm_idx_stat::17.454*1.00] distinct minimizers: 44056556 (79.15% are singletons); average occurrences: 1.918; average spacing: 2.942
[M::worker_pipeline::146.993*1.00] mapped 25483 sequences
[M::main] Version: 2.10-r763-dirty
[M::main] CMD: minimap2 -x ava-ont -t 1 MAP006-1_2D_pass.fasta MAP006-1_2D_pass.fasta
[M::main] Real time: 147.003 sec; CPU: 147.000 sec


Miniasm, we take these overlapping reads and assemble them into contigs.  Need the raw reads from poretools and the overlap file from minimap2 (I don't really understand what this means so once again gonna use a paragraph from brandon).  (because this outputs a .gfa file, we're piping the output into awk and using that to convert it to fasta format, then writing it to asm.fa)

A contig is a combination of several shorter reads that overlap to make one continuous sequence.  This "contiguous" segment of DNA is thought to be a true representation of a piece of the organism's genome.

In [23]:
!miniasm -f MAP006-1_2D_pass.fasta overlap.paf | awk '/^S/{print ">"$2"\n"$3}' | fold > asm.fa

[M::main] ===> Step 1: reading read mappings <===
[M::ma_hit_read::1.251*1.00] read 1202285 hits; stored 1888486 hits and 24307 sequences (241394465 bp)
[M::main] ===> Step 2: 1-pass (crude) read selection <===
[M::ma_hit_sub::1.570*1.00] 24211 query sequences remain after sub
[M::ma_hit_cut::1.602*1.00] 1879366 hits remain after cut
[M::ma_hit_flt::1.635*1.00] 1836573 hits remain after filtering; crude coverage after filtering: 45.23
[M::main] ===> Step 3: 2-pass (fine) read selection <===
[M::ma_hit_sub::1.751*1.00] 24203 query sequences remain after sub
[M::ma_hit_cut::1.782*1.00] 1836216 hits remain after cut
[M::ma_hit_contained::1.819*1.00] 1580 sequences and 15813 hits remain after containment removal
[M::main] ===> Step 4: graph cleaning <===
[M::ma_sg_gen] read 15464 arcs
[M::main] ===> Step 4.1: transitive reduction <===
[M::asg_arc_del_trans] transitively reduced 11947 arcs
[M::asg_arc_del_multi] removed 0 multi-arcs
[M::asg_arc_del_asymm] removed 157 asymmetric arcs
[M::mai

In order to determine which species are found  within our samples we will first need to find and/or build a database of species. We will get the sequence for each genome from the Ensemble Genome Database and the SGD. For this example we will simply pull 5 genomes. However, we could easily build a larger database. 

Note: If we wanted to build a database of  1000 genomes each 0.5Gb in size it would require a hard drive with 500Gb of space.  

This  is  accmplished with the wget cmd line tool. (Here the  mv tool is used to rename the yeast genome to contain the name  yeast and to relabel is extension .fasta to .fa)

In [14]:
# Retrieve genome sequences for each species
!wget ftp://ftp.ensemblgenomes.org/pub/bacteria/release-33/fasta/bacteria_9_collection/escherichia_coli_0_1304/dna/Escherichia_coli_0_1304.GCA_000303235_1.dna.nonchromosomal.fa.gz
!wget ftp://ftp.ensemblgenomes.org/pub/release-38/bacteria//fasta/bacteria_15_collection/_clostridium_asparagiforme_dsm_15981/dna/_clostridium_asparagiforme_dsm_15981.ASM15807v1.dna.nonchromosomal.fa.gz
!wget ftp://ftp.ensemblgenomes.org/pub/release-38/bacteria//fasta/bacteria_176_collection/_bacillus_aminovorans/dna/_bacillus_aminovorans.ASM164324v1.dna.nonchromosomal.fa.gz
!wget https://downloads.yeastgenome.org/sequence/S288C_reference/orf_dna/orf_genomic_1000_all.fasta.gz 
!mv  orf_genomic_1000_all.fasta.gz yeast_genomic_1000_all.fa.gz 

--2018-04-04 23:45:30--  ftp://ftp.ensemblgenomes.org/pub/bacteria/release-33/fasta/bacteria_9_collection/escherichia_coli_0_1304/dna/Escherichia_coli_0_1304.GCA_000303235_1.dna.nonchromosomal.fa.gz
           => ‘Escherichia_coli_0_1304.GCA_000303235_1.dna.nonchromosomal.fa.gz.1’
Resolving ftp.ensemblgenomes.org (ftp.ensemblgenomes.org)... 193.62.197.94
Connecting to ftp.ensemblgenomes.org (ftp.ensemblgenomes.org)|193.62.197.94|:21... connected.
Logging in as anonymous ... Logged in!
==> SYST ... done.    ==> PWD ... done.
==> TYPE I ... done.  ==> CWD (1) /pub/bacteria/release-33/fasta/bacteria_9_collection/escherichia_coli_0_1304/dna ... done.
==> SIZE Escherichia_coli_0_1304.GCA_000303235_1.dna.nonchromosomal.fa.gz ... 1679590
==> PASV ... done.    ==> RETR Escherichia_coli_0_1304.GCA_000303235_1.dna.nonchromosomal.fa.gz ... done.
Length: 1679590 (1.6M) (unauthoritative)

Escherichia_coli_0_ 100%[===================>]   1.60M  1.22MB/s    in 1.3s    

2018-04-04 23:45:34 (1.22 MB/s

There are many techniques to identify which species you have sequenced. One example is to try to align each of our contigs to each reference  in our database. However that is extreamly computationally intensive. Instead here we will use a computationally light application called MASH which calculates an  estimate  for  the  distance between our querry sequence and each reference.

There are many techniques to identify which species you have sequenced. One approach would be to try to align each of our contigs to each reference in our database. However, that would be extremely computationally intensive. Instead here we will use a computationally light application called MASH, which calculates an estimate for the distance between our query sequence and each reference.  In this context, distance refers to how similar or different two genomes are.

# Mash for taxonomy classification
First we must form a sketch from each of our reference genomes. A sketch is a minified representation of a reference. This contains less information than a whole genome sequence, but still allows us to calculate the distance between each of the references and our query.

We have already downloaded the sequences for each sample species described above. Now we must build the sketch of each reference using the command sketch from the mash tool.

First we  must form a sketch from each of our reference genomes. A sketch is a minified representation of a reference. However,it maintains the protperty that we  can still  calculate  the distance between each of the  references and our quarry.

We have already downloaded the sequences for each sample speceis described above. Now we must build the sketch of each reference this is done  using  the command sketch. 

Here we use  the -o options to name the sketch output file.

You should be able to see these four files in your directory now.

In [20]:
# The -o option allows us to name the sketch output file.
!mash sketch *.fa.gz -o reference

Sketching Escherichia_coli_0_1304.GCA_000303235_1.dna.nonchromosomal.fa.gz...
Sketching _bacillus_aminovorans.ASM164324v1.dna.nonchromosomal.fa.gz...
Sketching _clostridium_asparagiforme_dsm_15981.ASM15807v1.dna.nonchromosomal.fa.gz...
Sketching yeast_genomic_1000_all.fa.gz...
Writing to reference.msh...


Now we have our database of references in one file, reference.msh.  We want to take our assembly from before and compare it to the reference database.

When we assembled our contigs using miniasm, it tried to piece together all of the individual strands of DNA sequenced by the Nanopore.  If more than one species was present or the assembler couldn't figure out where to place small pieces of the organism's genome, then the assembly will be made of multiple contigs.  If we sequenced human DNA, we would naturally get many contigs, one for each chromosome.  In practice, it's hard to piece together the whole genome so even within one chromosome some gaps may remain unfilled.

To build a sketch for each contig we must us the -i option of the sketch command. This option tells mash to treat each contig as its own seperate entity. Here we again use the -o option as described above.

In [24]:
# Note: if there's no star on the left side of the code block ("In [*]:") then you know it's done running.
!mash sketch -i asm.fa -o contigs

Sketching asm.fa...
Writing to contigs.msh...


Now we are prepared to compute the distances between each contig and each reference. This is accomplished using the dist command. The '>' signifies to take the output from the following program and create a file called distance.tab.

In [25]:
!mash dist reference.msh contigs.msh > distance.tab

We can use the head command to look at the first five lines of this distance file. This file is organized into the following columns: [reference-ID, query-ID, distance, p-value, shared-hashes]

In [26]:
!head distance.tab

Escherichia_coli_0_1304.GCA_000303235_1.dna.nonchromosomal.fa.gz	utg000001l	0.0843302	0	93/1000
_bacillus_aminovorans.ASM164324v1.dna.nonchromosomal.fa.gz	utg000001l	1	1	0/1000
_clostridium_asparagiforme_dsm_15981.ASM15807v1.dna.nonchromosomal.fa.gz	utg000001l	1	1	0/1000
yeast_genomic_1000_all.fa.gz	utg000001l	1	1	0/1000


Now using a little python wizardry we  can find  the  most likely genome from our Database that represents each of our contigs from our  initial assebmly.  

What we want to do is determine which species has the miminum  distance  from  each contig. This is accmplished below by building a dictionary that  associates a contig with its closest reference throwing away all others.

If you do not know python dont be intimidated!! You can ninja too.

In [27]:
## A Shirt Code snippit that gets the closest Reference  for each Contig
import collections                                                             ##This line loads some extra python features for use
Reference = collections.namedtuple('Reference', ['name', 'distance','pvalue']) ##This creates a dummy "class/namespace" That has attributes for each componants of a distance measurement.
contigs   = collections.defaultdict(lambda : Reference(None, 1, 1)  )          ##A mapping between contigs and Reference_distances. By default each Contig is given the "null" reference.
with open("distance.tab")  as infile:                                          ##Opens a file for  reading this file is  a  list  of  lines.
    for reference, contig, distance, pvalue, matches in map(str.split,infile): ##this callsthe  string  split  function  on each line and  then upacks the splitline into 4 named columns.
        candidate = Reference(str(reference), float(distance), float(pvalue))  ##Creates a  Reference object for each contig  reference pair
        if contigs[contig].distance > candidate.distance:                      ##Compares each reference to the best reference.
                contigs[contig]  = candidate                                   ##Keeps the best  reference


In [31]:
##Print out the best Reference for each Contig
for contig, reference in contigs.items():            # Goes Through each contig  reference pair
    print("P-value   = ", reference.pvalue)
    print("Reference = ", reference.name)
    print("contig    = ",          contig)  # Prints the conig referenceand pvalue The best one!
    print()
##Can Print every Uniq taxa  found in the contig list!
# for taxa in set(reference.name  for reference  in  contigs.values()): #Iterates through every uniue taxa
#     print(taxa)                                                       #Prints out  each  unique taxa.

P-value   =  0.0
Reference =  Escherichia_coli_0_1304.GCA_000303235_1.dna.nonchromosomal.fa.gz
contig    =  utg000001l



# Using MetaGenomeMark to Describe Gene Information
- meta genome mark will take the fasta file of the assembled genome
- tell you what genes are in the genome


genomeOutputAssembled.fa - input file from Canu

In [32]:
!pwd
!ls /work/MetaGeneMark_linux_64/mgm

/work/data
gmhmmp	INSTALL  LICENSE  MetaGeneMark_v1.mod  README.MetaGeneMark


## This will obtain a GFF file

In [33]:
!gmhmmp -a -r -f G -d -m ../MetaGeneMark_linux_64/mgm/MetaGeneMark_v1.mod -o sequence.gff asm.fa


### Using this gff file, we can learn what genes are in your sample!

A GFF file has the columns: 
- seqname - name of the chromosome or scaffold
- source - name of the program that generated this feature: GeneMark.hmm
- feature - name of Gene, Variation, or Similarity
- start - Start position of the feature, with sequence numbering starting at 1.
- end - End position of the feature, with sequence numbering starting at 1.
- score - A floating point value.
- strand - defined as + (forward) or - (reverse).
- frame - One of '0', '1' or '2'. '0' indicates that the first base of the feature is the first base of a codon, '1' that the second base is the first base of a codon, and so on..
- attribute - A semicolon-separated list of tag-value pairs, providing additional information about each feature.

### Quick look at GFF file

In [34]:
!head -20 sequence.gff

##gff-version 2
##source-version GeneMark.hmm_PROKARYOTIC 3.38
##date: Thu Apr  5 00:00:46 2018
# Sequence file name: asm.fa
# Model file name: ../MetaGeneMark_linux_64/mgm/MetaGeneMark_v1.mod
# RBS: true

# Model information: Heuristic_model_for_genetic_code_11_and_GC_51

utg000001l	GeneMark.hmm	CDS	1	72	2.248856	+	0	gene_id=1
utg000001l	GeneMark.hmm	CDS	1984	2214	0.337658	+	0	gene_id=2
utg000001l	GeneMark.hmm	CDS	2289	2411	-2.242110	+	0	gene_id=3
utg000001l	GeneMark.hmm	CDS	3078	3404	-8.565622	+	0	gene_id=4
utg000001l	GeneMark.hmm	CDS	3420	3665	5.495138	+	0	gene_id=5
utg000001l	GeneMark.hmm	CDS	11382	11618	0.812308	-	0	gene_id=6
utg000001l	GeneMark.hmm	CDS	12156	12431	-0.839825	-	0	gene_id=7
utg000001l	GeneMark.hmm	CDS	12557	12790	2.258979	+	0	gene_id=8
utg000001l	GeneMark.hmm	CDS	13044	13190	6.278825	+	0	gene_id=9
utg000001l	GeneMark.hmm	CDS	15513	15620	-1.745413	-	0	gene_id=10
utg000001l	GeneMark.hmm	CDS	15624	15899	-8.697843	-	0	gene_id=11


### Goal: Getting FASTA files of all contigs listed 
This is your assembly file

In [35]:
!head asm.fa

>utg000001l
TTGCTCGGTTTTATTACTTTAGGCATTTATACTCCGCTGGAAGCGTGTGACCTGCTCAAAATAATTGCATGAGTTGCCCA
TCGATTGTAAGCTCTATTGAGCACTGCTCATTAATATACTTCTGGGTTCCTCAGTTCCAGTTGTTTTGCATAGTGATCAG
CCTCTCTGAGGGTGAAATAATCCCGTTCAGCGGTGTCTGCCAGTCGGGGGGAGGCTGCATTATCACGCCGGAGGCTGCGG
CTTCACGCATGACTGACAGACTGCTTTGATGTGCAACCGACGACCAAGAGCGGCAGCAACATCATCACGCAGAGCATCAT
TTTCAGCTTCGCATCAGCTAACTCCTTCGTGTATTTTGCAGCGACGCAGCAACATCACGCTGACGCATCTGCATGTCAGT
AATTGCCGCGTTCGCTAGCTTTTGCCAGTTCTCTCTGGCATTTTGTCGCCTGGACTTTGTAGGCGATTGCGTTATCACAC
GGTAATGATTGACCGCCCATGACAGGCTGACGATGATGCAGATAATCAGAGCGGATATAATCGCGGTTACTCTGCTCACT
GTTGCCCCCACAAACAGACTTCACGCTCAATCTCACGACGAGTCATCAGGCCTTTCCCATTATTGCTTACCGCCAGCGTA
TGTCCAGCGACGCAGCTGATGGATGCGCCTTTGATATCGCCCTGGTTTATTTTGCGAAGAAGCGTCGATGTTCTAAATTG


### The GFF file has unecessary headers that we don't need.
Run this sed command in unix to remove them

In [36]:
!sed -i -e 1,9d sequence.gff

This is our current GFF file, followed how many contigs found

In [37]:
!head -20 sequence.gff
!wc -l sequence.gff

utg000001l	GeneMark.hmm	CDS	1	72	2.248856	+	0	gene_id=1
utg000001l	GeneMark.hmm	CDS	1984	2214	0.337658	+	0	gene_id=2
utg000001l	GeneMark.hmm	CDS	2289	2411	-2.242110	+	0	gene_id=3
utg000001l	GeneMark.hmm	CDS	3078	3404	-8.565622	+	0	gene_id=4
utg000001l	GeneMark.hmm	CDS	3420	3665	5.495138	+	0	gene_id=5
utg000001l	GeneMark.hmm	CDS	11382	11618	0.812308	-	0	gene_id=6
utg000001l	GeneMark.hmm	CDS	12156	12431	-0.839825	-	0	gene_id=7
utg000001l	GeneMark.hmm	CDS	12557	12790	2.258979	+	0	gene_id=8
utg000001l	GeneMark.hmm	CDS	13044	13190	6.278825	+	0	gene_id=9
utg000001l	GeneMark.hmm	CDS	15513	15620	-1.745413	-	0	gene_id=10
utg000001l	GeneMark.hmm	CDS	15624	15899	-8.697843	-	0	gene_id=11
utg000001l	GeneMark.hmm	CDS	16628	16783	-4.989063	+	0	gene_id=12
utg000001l	GeneMark.hmm	CDS	17340	17516	-4.470129	-	0	gene_id=13
utg000001l	GeneMark.hmm	CDS	20076	20378	-6.331264	+	0	gene_id=14
utg000001l	GeneMark.hmm	CDS	21634	21828	4.955708	+	0	gene_id=15
utg000001l	GeneMark.hmm	CDS	23066	23227	0.340974	-	0	gen

### Run this to slice the FASTA sequences from the assembly from the GFF start and stop indecies

In [38]:
# This python script will get the start and stop indexes from the GFF 
# and get FASTA sequences from the assembly 

import csv

nameOfContig = list()
startIndexList = list()
stopIndexList = list()
# get start and stop indexes in the GFF file
with open("sequence.gff") as tsv:
    for line in csv.reader(tsv, dialect="excel-tab"): #You can also use delimiter="\t" rather than giving a dialect.
        if len(line) > 1:
            nameOfContig.append(""+str(line[2:3][0])+str(line[3:4][0])+"-"+str(line[4:5][0]))
            startIndexList.append(line[3:4])
            stopIndexList.append(line[4:5])
startAndStopList = list(zip(nameOfContig,startIndexList,stopIndexList))

# Use BioPython to assemble output FASTA file
from Bio import SeqIO
sequences = list()
for record in SeqIO.parse("asm.fa", "fasta"):
    print("This is the header for your assembly fasta: "+record.id)
    for name,start,stop in startAndStopList :
        if start != [] and stop != [] :
            sequences.append(record.seq[int(start[0]):int(stop[0])])
fastaList = list(zip(nameOfContig, sequences))
with open("annotatedGene.fa", "w") as output_handle:
    for name, seq in fastaList:
        fasta_format_string = ">"+name+"\n%s\n" % seq
        output_handle.write(fasta_format_string)

# Get the largest FASTA sequence
maxFasta = max(fastaList, key=lambda x: len(x[1]))
fasta_format_string = ">"+str(maxFasta[0])+"\n%s\n" % str(maxFasta[1])
print(fasta_format_string)

# Blastn the largest sequence
from Bio.Blast import NCBIWWW
result_handle = NCBIWWW.qblast("blastx", "nr", str(maxFasta[1]))



This is the header for your assembly fasta: utg000001l
>CDS3087214-3088254
TGGCGGATTACCACGTGCGTCGCGAAGAACGAGCAGATGGTGAGGCGCACTGGCAGAAAGAGATGACGCTACCGCGATGCGCTCTCTGTTTTGTTGCGCGGGGATGCGAGGGTGCTGCCGGTGGTGAACGACATGCAGGGCCAGCCTTGCGGCACGCTGCATTTGGAGCGTACTGCTGGTGGAGGCGTAAGCGTATGGCGACAGATGGTCTGGCCCCCAATCCGCTGTTCTGGCTCATTGCTCTGTTTGTGGCGCTCTTGATTTTCTGGCTGCCTTACAGCCAGCCGCTAGTTTGCTGCCTTGTTCCCACAACTGTGCCACGACCCGTTTATCAGCTACAAGAAGTTTTCAGCATCACTGGGGTGAGCCATTTCTGGCTGTGGGAATTCGCGTTTGTTGCGGTAGACTATTGGCTAACTGCCCCCGAATTGCGTCAAGCCCGTGGGCCGCGCCAAGATTTCGCCCACTGGTGGAGGAGAAACTATTGCCGCCGTGTTGGACAGACTATGTTCCGCCCGCGCATTCCTCGGTGCTGGCAAATCGCCCGTTCCGGTGATCGGCTGTTAGTCTGCAACCAGCGATTGCCGCGCCTTGATCACCTTTACGGTGTGCTCTGCCCGTCATGCGGCAGATGTTACAAGTCAAACCATAGATTGGTGCACAGAGCGTACGGAAGTTACTGGAAGGTATGGGAATCAGTCGTGATCCGAGTGCGTAAGTCGAGCTACCGCTGGCGGCTCCGGTGTAATTCTGGCGGCGTGCGAGAATTCATTATTAACATGGTACGGCGGGACGTACGCCTCAACGGTAGGGGCCCGCTAAAGCGCTGGGTACGCCCATCATCATCGGGCTGCGGATTTAATACCGCGTATGTGATCGGAGGGGCGTTACCGGGTGGCACTGAGGCGATCATCGCAACCGCCTG

### Annotated FASTA file is in /data
- This created a fasta of all the gffs, but theres alot!
- Copy the fasta sequence printed out from the last command and use the BLAST website to find hits to species!